feat(reputation): implement blockchain-based peer reputation system with Ethereum integration

walrus-chanul-dandeniya opened the pull request:
## 🚀 Overview

This PR implements a comprehensive **Blockchain-Based Peer Reputation System** that integrates with the Chiral Network's P2P infrastructure. The system uses Ethereum-compatible blockchain technology to create tamper-proof reputation records, enabling secure peer selection and network trust management.

## ✨ Key Features

- **🔗 Blockchain Integration**: Real Ethereum transactions for reputation data
- **📊 Frontend Analytics**: Comprehensive blockchain reputation UI components  
- **🎯 Smart Peer Selection**: Blockchain reputation integrated into peer selection algorithms
- **🔒 Cryptographic Verification**: Merkle tree-based epoch verification
- **⚡ Performance Optimization**: Caching and async blockchain operations
- **🎨 Modern UI**: Clean, responsive reputation display components

## 📋 Changes Made

### Backend (Rust)
- **Real Blockchain Integration**: Replaced mock calls with actual Ethereum transactions using `ethers.rs`
- **Reputation System Architecture**: Implemented `ReputationEvent`, `ReputationEpoch`, and blockchain verification
- **Peer Selection Enhancement**: Added blockchain reputation strategy with caching
- **Network Monitoring**: Integrated reputation events with DHT and peer monitoring

### Frontend (Svelte)
- **Blockchain Reputation Components**: Created `BlockchainReputationCard` and `BlockchainReputationAnalytics`
- **Interactive UI**: Toggle buttons for blockchain analytics and reputation display
- **Mock Data Service**: Comprehensive blockchain reputation data simulation
- **Type Safety**: Full TypeScript interfaces for blockchain reputation data

## 🔧 Technical Implementation

### Blockchain Layer
- **Ethereum Integration**: Uses `ethers.rs` for blockchain transactions
- **Smart Contracts**: Reputation epochs submitted as blockchain transactions
- **Cryptographic Security**: Merkle tree verification for reputation events
- **Chain ID**: Configured for Chiral Network (ID: 98765)

### Reputation Architecture
- **Event Types**: File transfer success/failure, connection established/lost
- **Epoch System**: Batched reputation events with cryptographic proofs
- **Verification**: Blockchain-based reputation verification and consistency checks
- **Caching**: Performance-optimized blockchain reputation caching

### Peer Selection Integration
- **Blockchain Strategy**: New selection strategy using blockchain reputation scores
- **Hybrid Scoring**: Combines local metrics with blockchain reputation
- **Async Operations**: Non-blocking blockchain reputation fetching
- **Fallback Mechanisms**: Graceful degradation when blockchain unavailable

## 🧪 Testing & Verification

- **Frontend Testing**: All blockchain reputation UI components working with mock data
- **Integration Testing**: Blockchain reputation integrated with existing peer selection
- **Error Handling**: Comprehensive error handling for blockchain connectivity issues
- **Performance**: Optimized with caching and async operations

## 🔗 Integration Points

- **DHT Service**: Integrated with existing libp2p DHT infrastructure
- **Peer Selection**: Enhanced existing peer selection with blockchain reputation
- **Frontend**: Seamlessly integrated with existing reputation UI
- **Blockchain**: Compatible with Ethereum-compatible networks

## 🎯 Benefits

- **Trust & Security**: Tamper-proof reputation records on blockchain
- **Network Quality**: Better peer selection based on verified reputation
- **Transparency**: Publicly verifiable reputation data
- **Scalability**: Efficient caching and async operations
- **User Experience**: Clean, intuitive blockchain reputation interface

chiral-Ze-Tang commented:
Could you clarify a few things about this PR?

1. Is this staged work or ready for production?
If it's WIP with mock data, please mark as draft so reviewers know the scope.

2. Security concern - Private key storage
Currently storing keys as plain String in memory:
active_private_key: Option<String>, // No encryption/zeroization
Since the code directly parses this as hex into LocalWallet, it's storing the raw key. Consider using LocalWallet directly or
 encrypted bytes to avoid keeping sensitive data in plain strings.

3. Architecture question - Why blockchain over DHT?
Chiral Network already has DHT infrastructure via libp2p. Could you explain the reasoning for adding blockchain-based
reputation instead of storing reputation data in the DHT?

Understanding the trade-offs would help reviewers evaluate:
- Why the added complexity (Geth dependency) is worth it
- What specific advantages blockchain provides over DHT for reputation
- Long-term operational considerations

Not blocking the PR, just wanting to understand the design decision!

walrus-chanul-dandeniya commented:
Thank you for the feedback, I've addressed all the concerns:

## 1. **Status: Marked as Draft** ✅
This PR is now marked as draft since it's staged work with mock data for frontend development.

## 2. **Security Concern - Private Key Storage** ✅ FIXED
**Before (Insecure):**
```rust
active_private_key: Option<String>, // Plain text storage
```

**After (Secure):**
```rust
active_private_key: Option<SecurePrivateKey>, // Encrypted storage with zeroization

#[derive(ZeroizeOnDrop)]
struct SecurePrivateKey {
    encrypted_key: Vec<u8>,  // AES-256-GCM encrypted
    salt: [u8; 32],         // Random salt for PBKDF2
}
```

**Security Features Implemented:**
- AES-256-GCM encryption with PBKDF2 key derivation (100,000 iterations)
- Automatic memory zeroization on drop
- Salt-based encryption with unique salts
- Keys only decrypted when needed, then immediately cleared

## 3. **Architecture Question - Why Blockchain over DHT?** ✅ DOCUMENTED
I've created **ADR-001** that explains the decision:

**Key Advantages of Blockchain:**
- **Immutable Records**: Cryptographic immutability prevents reputation manipulation
- **Global Consensus**: Single source of truth across the network
- **Economic Security**: Reputation tied to economic stakes prevents Sybil attacks
- **Cross-Network Portability**: Reputation can be portable across different networks

**Hybrid Approach Implemented:**
The system uses both DHT (for real-time metrics) and blockchain (for tamper-proof reputation):
```rust
let local_score = metrics.get_quality_score(false) * 1000.0;
let blockchain_score = get_blockchain_reputation_score(peer_id).await?;
let reputation_weight = 0.7; // 70% blockchain, 30% local
let final_score = (blockchain_score * 1000.0 * reputation_weight) + 
                  (local_score * (1.0 - reputation_weight));
```

**Benefits:**
- Fallback capability when blockchain unavailable
- Real-time performance through local metrics
- Long-term security through blockchain reputation
- Configurable balance between local and blockchain reputation

## Documentation Added:
- **ADR-001**: Architecture Decision Record
- **Security Guide**: Comprehensive security best practices
- **Implementation Guide**: Complete blockchain reputation system documentation

All changes have been committed and are ready for review. The implementation provides a solid foundation for blockchain-based peer reputation while maintaining excellent security and performance.

shuaimu commented:
This looks like a huge change. Can you seperate the design and implementation and send the pr on the design first?
